README.rst
=====
td.sh
=====

  Converting seconds to human readable time duration.

**td.sh** project has two implementations:

1. C:

   The C implementation include

   * library (``libtd``),
   * command-line program (``td``), and
   * Bash loadable (``td-bash``), loaded as ``td`` shell builtin.

2. Bash script (``td.sh``)

.. contents:: **Contents**
   :local:


Installation
============

When installing from Git repository, run the following command to initialize
required files:

.. code:: sh

  $ autoreconf --install

By default, to install to ``/usr/local``, run:

.. code:: sh

  $ ./configure && make && make install

Or to ``/usr``:

.. code:: sh

  $ ./configure prefix=/usr && make && make install

Or to your home:

.. code:: sh

  $ ./configure prefix=$HOME/.local && make && make install

To uninstall, use ``uninstall`` target.

Options
-------

``--disable-bash``
  do not build Bash implementation

``--disable-c``
  do not build C implementation

``--enable-bash-loadable=BASH_HEADER_PATH``
  build Bash loadable

  The loadable is installed as ``prefix/bin/td.bash``, it can be enabled by, as
  long as ``td.bash`` is in ``$PATH``:

  .. code:: sh

    $ enable -f "$(which td.bash)" td
    $ td 123

``--disable-python``
  do not build Python bindings

  ``--disable-python2``
    do not build Python 2 binding

  ``--disable-python3``
    do not build Python 3 binding

``--without-check``
  do not run Check unittest


Usage
=====

The basic usage:

.. code:: sh

  $ td.sh 123
  2 minutes 3 seconds

You can source the Bash script to have ``print_td`` function for better performance:

.. code:: sh

  $ source td.sh
  $ print_td 123
  2 minutes 3 seconds
  $ ./td.sh 1 2 60 61
  1 second
  2 seconds
  1 minute
  1 minute 1 second
  $ ./td.sh -p -P 1 2 60 61
   1 second 
   2 seconds
   1 minute 
   1 minute   1 second 
  $ ./td.sh -p -P -a 1 2 60 61
   0 days  0 hours  0 minutes  1 second 
   0 days  0 hours  0 minutes  2 seconds
   0 days  0 hours  1 minute   0 seconds
   0 days  0 hours  1 minute   1 second 

Options
-------

``-a``
  prints all numbers and units even the numbers are zeros.

``-p[X]``: number padding
  prints out like ``_1 second`` (``_`` denotes the space) or ``-p0`` for ``01 second`` or ``07 seconds``, default is a space .

``-P``: unit string padding
  prints out like ``1 minute  12 seconds``, note there is additional space after ``minute``.


Tests
=====

The unittest is done by Check_ and Autotest_:

.. code:: sh

  $ make check

.. _Check: http://check.sourceforge.net/
.. _Autotest: http://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/html_node/Using-Autotest.html#Using-Autotest


Benchmark
=========

.. code:: sh

  $ make benchmark

Here is a sample result of benchmarking:

+-------------+---------------------------------+
| conversions | benchmark                       |
+=============+=================================+
| 2,060       | ``td.sh`` function calls        |
+-------------+---------------------------------+
| 184         | ``td.sh`` script executions     |
+-------------+---------------------------------+
| 60          | ``td`` command executions       |
+-------------+---------------------------------+
| 24,092      | ``td.bash`` loadable executions |
+-------------+---------------------------------+
| 40          | Python 2 script executions      |
+-------------+---------------------------------+
| 16          | Python 3 script executions      |
+-------------+---------------------------------+


License
=======

This project is licensed under the MIT License, see COPYING_ for full text::

  Copyright (c) 2010, 2012, 2014 Yu-Jie Lin

.. _COPYING: COPYING


NEWS
===========================
NEWS for 2.0.0 (2014-08-31)
===========================

Changes since 1.0.3 (2014-07-06)
================================

	* convert into C
	* use GNU Autotools to build
	* build  C library
	* Python 2 and Python 3 bindings
	* Bash loadable extension

[libtd]
	* version 1:0:0

[td]

	* add -h for help message
	* add -V for version string

[td.bash]

	* add -h for help message
	* add -V for version string


td-test.sh
#!/bin/bash

TD="src/td"

# Performance test

trap 'kill $tmpcount_pid ; rm "$tmpcount" ; exit 130' INT

echo -n "Please wait for 5 seconds..."
tmpcount="$(mktemp)"
( trap exit TERM; while :; do $TD 180122; echo >> "$tmpcount"; done ) &>/dev/null &
tmpcount_pid=$!
sleep 5
kill $tmpcount_pid
echo -ne "\e[2K\e[0G$(bc <<< "$(wc -l < "$tmpcount") / 5") conversions per second.\n"
rm "$tmpcount"


tests/check_libtd.c
/*
 * libtd uniti test
 * Copyright (c) 2014 Yu-Jie Lin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do
 * so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <stdbool.h>
#include <stdlib.h>

#include <check.h>

#include "../lib/td.h"

#define \
TEST_SPRINT_TD(str, t, all, pad, ch, expect) \
  do {                                       \
    str[0] = '\0';                           \
    sprint_td(str, t, all, pad, ch);         \
    ck_assert_str_eq(str, expect);           \
  } while (false);

START_TEST (test_libtd)
{
  char str[64] = "";

  TEST_SPRINT_TD(str,     0, false, false, '\0', "0 seconds");
  TEST_SPRINT_TD(str,     1, false, false, '\0', "1 second");
  TEST_SPRINT_TD(str,    -1, false, false, '\0', "1 second");
  TEST_SPRINT_TD(str,     2, false, false, '\0', "2 seconds");
  TEST_SPRINT_TD(str,    60, false, false, '\0', "1 minute");
  TEST_SPRINT_TD(str,    61, false, false, '\0', "1 minute 1 second");
  TEST_SPRINT_TD(str,  3599, false, false, '\0', "59 minutes 59 seconds");
  TEST_SPRINT_TD(str,  3600, false, false, '\0', "1 hour");
  TEST_SPRINT_TD(str,  3601, false, false, '\0', "1 hour 1 second");
  TEST_SPRINT_TD(str,  3660, false, false, '\0', "1 hour 1 minute");
  TEST_SPRINT_TD(str,  3661, false, false, '\0', "1 hour 1 minute 1 second");
  TEST_SPRINT_TD(str, 86400, false, false, '\0', "1 day");

  TEST_SPRINT_TD(str,       86400 + 86400 - 1, false, false, '\0', "1 day 23 hours 59 minutes 59 seconds");
  TEST_SPRINT_TD(str,   2 * 86400 + 86400 - 1, false, false, '\0', "2 days 23 hours 59 minutes 59 seconds");
  TEST_SPRINT_TD(str, 366 * 86400         + 1, false, false, '\0', "366 days 1 second");

  TEST_SPRINT_TD(str,     1, false,  true, '\0', "1 second ");
  TEST_SPRINT_TD(str,    60, false,  true, '\0', "1 minute ");
  TEST_SPRINT_TD(str,    60, false,  true,  ' ', " 1 minute ");
  TEST_SPRINT_TD(str,    60, false,  true,  '0', "01 minute ");

  TEST_SPRINT_TD(str,     0,  true, false, '\0', "0 days 0 hours 0 minutes 0 seconds");

  TEST_SPRINT_TD(str,    60,  true, false,  ' ', " 0 days  0 hours  1 minute  0 seconds");
  TEST_SPRINT_TD(str,    60,  true,  true, '\0', "0 days 0 hours 1 minute  0 seconds");
  TEST_SPRINT_TD(str,    60,  true,  true,  ' ', " 0 days  0 hours  1 minute   0 seconds");
}
END_TEST

Suite *
libtd_suite (void)
{
  Suite *suite = suite_create("libtd");

  TCase *tc_libtd = tcase_create("libtd");
  tcase_add_test(tc_libtd, test_libtd);
  suite_add_tcase(suite, tc_libtd);

  return suite;
}

int
main (void)
{
  int number_failed;

  Suite *s = libtd_suite();
  SRunner *sr = srunner_create(s);
  srunner_run_all(sr, CK_NORMAL);
  number_failed = srunner_ntests_failed(sr);
  srunner_free(sr);

  return (number_failed == 0) ? EXIT_SUCCESS : EXIT_FAILURE;
}



tests/testsuite.td.py
import sys

import td


if '--' in sys.argv:
  durations = sys.argv[sys.argv.index('--') + 1:]
else:
  durations = filter(lambda n: '-' not in n, sys.argv[1:])
print_all_numbers = '-a' in sys.argv
pad_units = '-P' in sys.argv
pad_char = ' ' if '-p' in sys.argv else ('0' if '-p0' in sys.argv else '\000')


for d in durations:
  if '-s' in sys.argv:
    print(td.sprint_td(int(d), print_all_numbers, pad_units, pad_char))
  else:
    td.print_td(int(d), print_all_numbers, pad_units, pad_char)


td.sh
#!/bin/bash
# Converting seconds to human readable time duration.
# Copyright (c) 2010, 2012, 2014 Yu-Jie Lin
# 
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

usage() {
  echo "Usage: $(basename "$0") [options] [seconds[ seconds[ seconds[...]]]]

Options:

  -a     prints all number even they are zeros.
  -p[X]  padding for numbers. 'X' is the padding character, default is ' '.
  -P     padding for unit strings.
  -h     this help message.

Note:

  You can \`source $(basename "$0")\` then use \`print_td <seconds>\` in your script.
"
}

units=(second minute hour day)

print_td() {
  t=${1#-}
  nums=($(($t % 60)) $(($t / 60 % 60)) $(($t / 3600 % 24)) $(($t / 86400)))

  result=""
  for ((idx=${#units[@]}-1;idx>=0;idx--)) {
    ((nums[idx] == 0)) && [[ -z $TD_SH_PRINTS_ZEROS ]] && continue;
    # Handling Number Padding
    if (( nums[idx] < 10 )) && [[ ! -z $TD_SH_NUMB_PADDING ]]; then
      result="$result ${TD_SH_NUMB_PADDING}${nums[idx]}"
    else
      result="$result ${nums[idx]}"
    fi

    result="$result ${units[idx]}"
    # Handling Unit Padding
    if ((nums[idx] != 1)); then
      result="${result}s"
    elif [[ ! -z $TD_SH_UNIT_PADDING ]]; then
      result="${result} "
    else
      result="${result}"
    fi
    }
  [[ -z "$result" ]] && result="0 seconds"
  echo "${result# }"
  }

shopt -s extglob

TD_SH_PRINTS_ZEROS=
TD_SH_UNIT_PADDING=
TD_SH_NUMB_PADDING=


eval set -- $(getopt -- aPp::h "$@")
while (($#)); do
  case "$1" in
    -a)
      TD_SH_PRINTS_ZEROS="ON"
      shift
      ;;
    -P)
      TD_SH_UNIT_PADDING="ON"
      shift
      ;;
    -p)
      TD_SH_NUMB_PADDING=${2:- }
      shift 2
      ;;
    -h)
      usage
      ;;
    --)
      shift
      break
  esac
done

for t in $@; do
  print_td "$t"
done


lib/td.h
/***********************************************************************************/
/* Copyright (c) 2014 Yu-Jie Lin                                                   */
/*                                                                                 */
/* Permission is hereby granted, free of charge, to any person obtaining a copy of */
/* this software and associated documentation files (the "Software"), to deal in   */
/* the Software without restriction, including without limitation the rights to    */
/* use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies   */
/* of the Software, and to permit persons to whom the Software is furnished to do  */
/* so, subject to the following conditions:                                        */
/*                                                                                 */
/* The above copyright notice and this permission notice shall be included in all  */
/* copies or substantial portions of the Software.                                 */
/*                                                                                 */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR      */
/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,        */
/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE     */
/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER          */
/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,   */
/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE   */
/* SOFTWARE.                                                                       */
/***********************************************************************************/

#include <stdbool.h>

int sprint_td(char *str, long long t, bool print_all_numbers, bool pad_units, char pad_char);
void print_td(long long t, bool print_all_numbers, bool pad_units, char pad_char);


src/main.c
/***********************************************************************************/
/* td: main.c - the C of td.sh                                                     */
/* Copyright (c) 2014 Yu-Jie Lin                                                   */
/*                                                                                 */
/* Permission is hereby granted, free of charge, to any person obtaining a copy of */
/* this software and associated documentation files (the "Software"), to deal in   */
/* the Software without restriction, including without limitation the rights to    */
/* use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies   */
/* of the Software, and to permit persons to whom the Software is furnished to do  */
/* so, subject to the following conditions:                                        */
/*                                                                                 */
/* The above copyright notice and this permission notice shall be included in all  */
/* copies or substantial portions of the Software.                                 */
/*                                                                                 */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR      */
/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,        */
/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE     */
/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER          */
/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,   */
/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE   */
/* SOFTWARE.                                                                       */
/***********************************************************************************/

#include "../config.h"
#include <td.h>

#ifndef BASH_LOADABLE
#include <getopt.h>
#endif
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#ifdef BASH_LOADABLE
#include <builtins.h>
#include <shell.h>
#include <bashgetopt.h>
#endif

#ifndef BASH_LOADABLE
#define NAME    PACKAGE_NAME
#else
#define NAME    PACKAGE_NAME ".bash"
#endif
#define USAGE   PACKAGE_NAME " [OPTIONS] [--] [[duration] ...]"

#define PRINT_VERSION()  puts(NAME " " PACKAGE_VERSION)

#ifndef BASH_LOADABLE
#define GETOPT()  getopt(argc, argv, "hVaPp::")
#define OPTARG    optarg
#else
#define GETOPT()  internal_getopt(list, "hVaPp;")
#define OPTARG    list_optarg
#endif
#define URL_WEB   "https://github.com/livibetter/td.sh"
#define URL_BUG   "https://github.com/livibetter/td.sh/issues"


char *td_doc[] = {
  "Convert seconds to human readable time duration.",
  "",
  "Options:",
  "",
  "  -h         display this help message",
  "  -V         display version string",
  "  -a         print all numbers and units",
  "  -P         unit string padding",
  "  -p[X]      number padding using character X",
  (char *) NULL
};


void
print_help () {
  unsigned int i;

  puts("Usage: " USAGE);
  for (i = 0; i < sizeof(td_doc) / sizeof(char *); i++)
    if (td_doc[i])
      puts(td_doc[i]);

  printf("\n"\
         "Report bugs to <" URL_BUG ">\n"\
         "Home page: <" URL_WEB ">\n");
}


int
#ifndef BASH_LOADABLE
main(int argc, char *argv[])
#else
td_builtin (WORD_LIST *list)
#endif
{
  int i;
  int opt;
  bool print_all_numbers = false;
  bool pad_units = false;
  char pad_char = '\0';
  long long t;

#ifdef BASH_LOADABLE
  reset_internal_getopt();
#endif
  while ((opt = GETOPT()) != -1) {
    switch (opt) {
    case 'h':
      print_help();
      goto out;
    case 'V':
      PRINT_VERSION();
      goto out;
    case 'a':
      print_all_numbers = true;
      break;
    case 'P':
      pad_units = true;
      break;
    case 'p':
      if (!OPTARG)
        pad_char = ' ';
      else
        pad_char = OPTARG[0];
      break;
    default:
#ifndef BASH_LOADABLE
      fprintf(stderr, "Usage: " USAGE "\n");
      return EXIT_FAILURE;
#else
      return EX_USAGE;
#endif
    }
  }

#ifndef BASH_LOADABLE
  for (i = optind; i < argc; i++) {
    t = atoll(argv[i]);
#else
  for (list = loptend; list; list = list->next) {
    t = atoll(list->word->word);
#endif
    print_td(t, print_all_numbers, pad_units, pad_char);
  }

out:
#ifndef BASH_LOADABLE
  return EXIT_SUCCESS;
#else
  return EXECUTION_SUCCESS;
#endif
}


#ifdef BASH_LOADABLE
struct builtin td_struct = {
  "td",
  td_builtin,
  BUILTIN_ENABLED,
  td_doc,
  USAGE,
  0
};
#endif


lib/td.c
/***********************************************************************************/
/* td.c Converting seconds to human readable time duration.                        */
/* Copyright (c) 2014 Yu-Jie Lin                                                   */
/*                                                                                 */
/* Permission is hereby granted, free of charge, to any person obtaining a copy of */
/* this software and associated documentation files (the "Software"), to deal in   */
/* the Software without restriction, including without limitation the rights to    */
/* use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies   */
/* of the Software, and to permit persons to whom the Software is furnished to do  */
/* so, subject to the following conditions:                                        */
/*                                                                                 */
/* The above copyright notice and this permission notice shall be included in all  */
/* copies or substantial portions of the Software.                                 */
/*                                                                                 */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR      */
/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,        */
/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE     */
/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER          */
/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,   */
/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE   */
/* SOFTWARE.                                                                       */
/***********************************************************************************/

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *units[] = {"second", "minute", "hour", "day"};
int n_units = sizeof(units) / sizeof(char *);

int
sprint_td(char *str, long long t, bool print_all_numbers, bool pad_units, char pad_char)
{
  int i;
  long long n[] = {
    abs(t) % 60,
    abs(t) / 60 % 60,
    abs(t) / 3600 % 24,
    abs(t) / 86400
  };
  /* space between two components */
  bool need_space = false;

  if (t == 0 && !print_all_numbers) {
    sprintf(str, "0 seconds");
    return strlen(str);
  }

  for (i = n_units - 1; i >= 0; i--) {
    if (n[i] || print_all_numbers) {
      if (need_space) {
        strcat(str + strlen(str), " ");
        need_space = false;
      }
      if (n[i] < 10)
        sprintf(str + strlen(str), "%c", pad_char);
      if (n[i] == 1) {
        if (pad_units)
          sprintf(str + strlen(str), "%lld %s ", n[i], units[i]);
        else
          sprintf(str + strlen(str), "%lld %s", n[i], units[i]);
      } else
        sprintf(str + strlen(str), "%lld %ss", n[i], units[i]);
      need_space = true;
    }
  }

  return strlen(str);
}

void
print_td(long long t, bool print_all_numbers, bool pad_units, char pad_char)
{
  char str[64] = "";

  sprint_td(str, t, print_all_numbers, pad_units, pad_char);
  puts(str);
}
